src/mlpack/tests/ann_layer_test.cpp

```diff
@@ -1282,4 +1282,104 @@ BOOST_AUTO_TEST_CASE(SimpleMeanSquaredErrorLayerTest)
+
+  arma::mat input = dataset.submat(0, 0, dataset.n_rows-1, 2);
+
+  size_t numUnits = input.n_rows;
+
+  BatchNorm<> model(numUnits);
+  model.Reset();
+
+  // Non-Deteministic Forward Pass Test.
+  model.Deterministic() = false;
+  model.Forward(std::move(input), std::move(output));
+
+  arma::mat result;
+  result << 1.20240722e+00 << 5.33976074e-15 << -1.20240722e+00 << arma::endr
+         << 1.28267074e+00 << -1.12233689e+00 << -1.60333842e-01 << arma::endr
+         << 5.87220220e-01 << 5.87220220e-01 << -1.17444044e+00 << arma::endr
+         << 0.00000000e+00 << 0.00000000e+00 << 0.00000000e+00 << arma::endr;
+
+  CheckMatrices(output, result);
+  result.clear();
+
+  // Backward Pass Test.
+  arma::mat gy;
+  gy << 0.8402 << 0.9116 << 0.2778 << arma::endr
+     << 0.3944 << 0.1976 << 0.5540 << arma::endr
+     << 0.7831 << 0.3352 << 0.4774 << arma::endr
+     << 0.7984 << 0.7682 << 0.6289 << arma::endr;
+
+  model.Backward(std::move(input), std::move(gy), std::move(output));
+
+  result << -0.64550918 << 1.41322929 << -0.76772011 << arma::endr
+         << -0.34197354 << -0.5355513 << 0.87752484 << arma::endr
+         <<  4.09422086 << -3.79625723 << -0.29796364 << arma::endr
+         <<  2.10502283 << 1.15001498 << -3.2550378  << arma::endr;
+
+  CheckMatrices(output, result);
+  result.clear();
+
+  // Gradient Test.
+  model.Gradient(std::move(input), std::move(gy), std::move(output));
+
+  result << 0.67623382 << arma::endr
+         << 0.19528662 << arma::endr
+         << 0.09601051 << arma::endr
+         << 0.00000000 << arma::endr
+         << 2.0296 << arma::endr
+         << 1.146 << arma::endr
+         << 1.5957 << arma::endr
+         << 2.1955 << arma::endr;
+
+  CheckMatrices(output, result);
+  result.clear();
+
+  // Deterministic Forward Pass test.
+  input = dataset.submat(0, 3, dataset.n_rows-1, 5);
+  model.Forward(std::move(input), std::move(output));
+
+  input = dataset.submat(0, 6, dataset.n_rows-1, 8);
+  model.Forward(std::move(input), std::move(output));
+
+  output = model.TrainingMean();
+  result << 4.85555556 << arma::endr
+         << 3.33333333 << arma::endr
+         << 1.44444444 << arma::endr
+         << 0.23333333 << arma::endr;
+
+  CheckMatrices(output, result);
+  result.clear();
+
+  output = model.TrainingVariance();
+  result << 0.08469136 << arma::endr
+         << 0.08888889 << arma::endr
+         << 0.01135802 << arma::endr
+         << 0.00444444 << arma::endr;
+
+  CheckMatrices(output, result, 2e-4);
+  result.clear();
+
+  input = dataset.submat(0, 0, dataset.n_rows-1, 2);
+
+  model.Deterministic() = true;
+  model.Forward(std::move(input), std::move(output));
+
+  result << 0.83504842 <<  0.15182699 << -0.53139445 << arma::endr
+         << 0.55589881 << -1.11179762 << -0.44471905 << arma::endr
+         << -0.39980015 << -0.39980015 << -1.29935049 << arma::endr
+         << -0.45175395 << -0.45175395 << -0.45175395 << arma::endr;
+
+  CheckMatrices(output, result);
+}
```

Comment:
It might be a good idea to also add a test that uses the `CheckGradient()` function; this is a nice way to ensure that the implementation of the objective matches the implementation of the gradient.

Reply:
I have added it in the latest commit.